dGame/dUtilities/Loot.cpp

```diff
@@ -1,384 +1,385 @@
+    uniqueRarityIndices.erase(std::unique(uniqueRarityIndices.begin(), uniqueRarityIndices.end()), uniqueRarityIndices.end());
 
-	DropLoot(user, killedObject, result, minCoins, maxCoins);
-}
+    for (const uint32_t index : uniqueRarityIndices) {
+        std::vector<CDRarityTable> table = rarityTableTable->Query([index](const CDRarityTable& entry) { return entry.RarityTableIndex == index; });
 
-void Loot::GiveLoot(Entity* user, std::unordered_map<LOT, int32_t>& result) {
-	user = user->GetOwner(); // If the owner is overwritten, we collect that here
+        RarityTable rarityTable;
 
-	auto* inventoryComponent = user->GetComponent<InventoryComponent>();
+        for (const CDRarityTable& entry : table) {
+            RarityTableEntry rarity{entry.rarity, entry.randmax};
+            rarityTable.push_back(rarity);
+        }
 
-	if (inventoryComponent == nullptr) {
-		return;
-	}
+        // sort in descending order based on randMax
+        std::sort(rarityTable.begin(), rarityTable.end(), [](const RarityTableEntry& x, const RarityTableEntry& y) { return x.randMax > y.randMax; });
 
-	for (const auto& pair : result) {
-		inventoryComponent->AddItem(pair.first, pair.second);
-	}
-}
+        m_RarityTables.insert({index, rarityTable});
+    }
+
+    // ==============================
+    // Cache Loot Matrices
+    // ==============================
 
-void Loot::DropLoot(Entity* user, Entity* killedObject, std::unordered_map<LOT, int32_t>& result, uint32_t minCoins, uint32_t maxCoins) {
-	user = user->GetOwner(); // If the owner is overwritten, we collect that here
+    std::vector<uint32_t> uniqueMatrixIndices;
 
-	auto* inventoryComponent = user->GetComponent<InventoryComponent>();
+    for (const CDLootMatrix& matrix : lootMatrixTable->GetEntries()) {
+        uniqueMatrixIndices.push_back(matrix.LootMatrixIndex);
+    }
 
-	if (inventoryComponent == nullptr) {
-		return;
-	}
+    // filter out duplicates
+    std::sort(uniqueMatrixIndices.begin(), uniqueMatrixIndices.end());
+    uniqueMatrixIndices.erase(std::unique(uniqueMatrixIndices.begin(), uniqueMatrixIndices.end()), uniqueMatrixIndices.end());
 
-	const auto spawnPosition = killedObject->GetPosition();
-	
-	const auto source = killedObject->GetObjectID();
+    for (const uint32_t index : uniqueMatrixIndices) {
+        std::vector<CDLootMatrix> matrix = lootMatrixTable->Query([index](const CDLootMatrix& entry) { return entry.LootMatrixIndex == index; });
 
-	for (const auto& pair : result) {
-		for (int i = 0; i < pair.second; ++i) {
-			GameMessages::SendDropClientLoot(user, source, pair.first, 0, spawnPosition, 1);
-		}
-	}
+        LootMatrix lootMatrix;
 
-	uint32_t coins = (int)(minCoins + GeneralUtils::GenerateRandomNumber<float>(0, 1) * (maxCoins - minCoins));
+        for (const CDLootMatrix& entry : matrix) {
+            LootMatrixEntry matrixEntry{entry.LootTableIndex, entry.RarityTableIndex, entry.percent, entry.minToDrop, entry.maxToDrop};
```

Comment:
This will result in a lot of double storage. Consider using a ptr to the already cached table entry.

Reply:
I think the speed of not checking through them on add outweighs the few kb of lost memory